### PR TITLE
DOCK-2591: Sort "representative" version first by default

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR changes the webservice's version retrieval endpoints to allow no sort column to be specified, in which case, the "representative" version (typically the default version) appears first in the results, and other results are sorted in recently-modified order.  The rationale here is that now that pagination happens on the database side, the database (rather than the UI) has the necessary information to decide the default sort behavior, and should do so.  This is analogous to how our ES search results work.

If  a sort column is specified, the results are sorted as they are now, the representative version appears wherever it should, per the specified ordering.

Also, we change the secondary sort criteria to be the last modified date (and the tertiary sort criteria becomes the created date), to be consistent with our current "default" table sort order.

Right now, in the UI, the version table sorts by "lastModified" when the table loads.  So, to use this PR, we'll need to modify the UI, probably by changing the column "sort direction" controls to have three possible states: asc, desc, and none.  When the version table first appears, the controls will be set to "none", signaling the webservice to return the results in default order.  Again, this is similar to how the search result table works.

The UI PR will happen after https://github.com/dockstore/dockstore-ui2/pull/2023 merges.

~~I didn't add a test, because it was relatively labor/code intensive.   Probably easier to do it in the UI, and possibly more appropriate.  In the interim, if this change breaks, it'll be very visible.~~

**Review Instructions**
Query the `getWorkflowVersions` endpoint without a sort column and confirm that the default/representative version appears first in the results.

**Issue**
https://ucsc-cgl.atlassian.net/browse/DOCK-2591
#6019 

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 

- [x] Security and Privacy assessed

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by denis-yuen** [comment]
> So, to use this PR, we'll need to modify the UI, probably by changing the column "sort direction" controls to have three possible states: asc, desc, and none. When the version table first appears, the controls will be set to "none", signaling the webservice to return the results in default order.

For understanding, what were the trade-offs of sorting the default version to always show as first and leaving the rest?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java [hunk 1]
```diff
@@ -119,19 +121,17 @@ private List<Predicate> processQuery(String sortCol, String sortOrder, CriteriaB
...
-                query.orderBy(cb.desc(versionId));
-            } else {
-                query.orderBy(cb.asc(versionId));
-            }
+            sortExpression = cb.selectCase()
+                .when(cb.equal(versionId, representativeVersionId), 1)
+                .otherwise(0);
+        }
+        if ("desc".equalsIgnoreCase(sortOrder)) {
+            query.orderBy(cb.desc(sortExpression), cb.desc(lastModified), cb.desc(versionId));
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java:132]: i.e. can you sort by whether or not a version is the representative version before sortExpression, last modified, versionID and/or what are the trade-offs

**Review by denis-yuen — COMMENTED** [review]
questions

**Comment by svonworl** [comment]
> > If a sort column is specified, the results are sorted as they are now, the representative version appears wherever it should, per the specified ordering.
> 
> If I understand correctly and going back to #6019, this guarantees that the default version will be displayed if there is no sort specified. But it doesn't guarantee it if a sort is applied, however, since we can't currently link to a sorted view of the versions table, this seems "ok" since someone would need to intentionally click to sort and thus "push" the default version out of the 200 version window. Is that correct?

Yup, the default version will be at the top initially.   One possible philosophical position (which I currently subscribe to) is that when the user specifies a sort, the default version should move to the appropriate position per the specified sort.  Otherwise, the table contents are sorted "wrong" and don't entirely make sense.  Yes, this means that the default version may not be displayed in the first 200 results when we sort.

That's probably the behavior I would expect as a user.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java (hunk 1 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java:132]: These are the primary, secondary, and tertiary sort criteria.  That is, if `sortExpression` results in a tie, `lastModified` is used to break the tie, and is `lastModified` results in a tie, `versionId` is used to break the tie.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java [hunk 2]
```diff
@@ -450,10 +450,10 @@ public Workflow getWorkflow(@ApiParam(hidden = true) @Parameter(hidden = true, n
     @ApiResponse(responseCode = HttpStatus.SC_BAD_REQUEST + "", description = "Bad Request")
     public Set<WorkflowVersion> getWorkflowVersions(@Parameter(hidden = true, name = "user") @Auth User user,
         @Parameter(
-                name = "workflowId", description = "id of the worflow", required = true, in = ParameterIn.PATH) @PathParam("workflowId") Long workflowId,
+                name = "workflowId", description = "id of the workflow", required = true, in = ParameterIn.PATH) @PathParam("workflowId") Long workflowId,
         @QueryParam("limit") @Min(1) @Max(MAX_PAGINATION_LIMIT) @DefaultValue(PAGINATION_LIMIT) Integer limit,
         @QueryParam("offset") @Min(0) @DefaultValue("0") Integer offset,
-        @DefaultValue("lastModified") @QueryParam("sortCol") String sortCol,
+        @QueryParam("sortCol") String sortCol,
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:456]: Might be worth adding an openapi description of what happens if no sortCol is specified

**Review by kathy-t — COMMENTED** [review]
From the issue description:

> For a published workflow with >200 versions, if the default version is older than the 200 versions that we retrieve from the webservice, it is not displayed in the Versions tab and the version selected is not the default version.

It's not immediately clear to me from this PR, but just checking, does this PR address "the version selected is not the default version" component of the issue?

**Comment by svonworl** [comment]
> From the issue description:
> 
> > For a published workflow with >200 versions, if the default version is older than the 200 versions that we retrieve from the webservice, it is not displayed in the Versions tab and the version selected is not the default version.
> 
> It's not immediately clear to me from this PR, but just checking, does this PR address "the version selected is not the default version" component of the issue?

Probably not, but I haven't looked into the UI change yet.  IMHO, the current rows (numbering 200 or less) in the version table are a window into the full list of versions, sorted as appropriate, and if the currently-selected version (which, after this PR and some UI mods, will appear initially, when no sort is specified) isn't one of the rows, no row should be marked as selected.  Will likely need to tweak the UI a little to that happen, though.

**Comment by denis-yuen** [comment]
> That's probably the behavior I would expect as a user.

To clarify, I think I can live with this since it's not a user request.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java [hunk 3]
```diff
@@ -477,18 +477,18 @@ public Set<WorkflowVersion> getWorkflowVersions(@Parameter(hidden = true, name =
     @ApiResponse(responseCode = HttpStatus.SC_BAD_REQUEST + "", description = "Bad Request")
     public Set<WorkflowVersion> getPublicWorkflowVersions(
             @Parameter(
-                name = "workflowId", description = "id of the worflow", required = true, in = ParameterIn.PATH) @PathParam("workflowId") Long workflowId,
+                name = "workflowId", description = "id of the workflow", required = true, in = ParameterIn.PATH) @PathParam("workflowId") Long workflowId,
         @QueryParam("limit") @Min(1) @Max(MAX_PAGINATION_LIMIT) @DefaultValue(PAGINATION_LIMIT) Integer limit,
         @QueryParam("offset") @Min(0) @DefaultValue("0") Integer offset,
-        @DefaultValue("lastModified") @QueryParam("sortCol") String sortCol,
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:483]: Copying Kathy's comment about openapi description if no sortcol here.

**Review by coverbeck — APPROVED** [review]
The `determineRepresentativeVersion` can end up iterating through all versions in Java, but:

1. Only for entries without a default version
2. They don't all come back in the payload.

We can always optimize that later -- your implementation is slick.

**Review by denis-yuen — COMMENTED** [review]
Commenting to silence notification

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java (hunk 2 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java]: Added description.

**Comment by svonworl** [comment]
> > I didn't add a test, because it was relatively labor/code intensive. Probably easier to do it in the UI, and possibly more appropriate. In the interim, if this change breaks, it'll be very visible.
> 
> If testing the sorting is a pain, would like to at least lock down the "default version shows up first when unsorted results" aspect of things.

Added integration test that tests that default version is sorted first.

**Review by denis-yuen — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

**Pull request merged**